examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
+        )
+
+    def causal_attention_mask(self, batch_size, n_dest, n_src, dtype):
+        """Mask the upper half of the dot product matrix in self attention.
+
+        This prevents flow of information from future tokens to current token.
+        1's in the lower triangle, counting from the lower right corner.
+        """
+        i = tf.range(n_dest)[:, None]
+        j = tf.range(n_src)
+        m = i >= j - n_src + n_dest
+        mask = tf.cast(m, dtype)
+        mask = tf.reshape(mask, [1, n_dest, n_src])
+        mult = tf.concat(
+            [tf.expand_dims(batch_size, -1), tf.constant([1, 1], dtype=tf.int32)], 0
+        )
+        return tf.tile(mask, mult)
+
+    def call(self, enc_out, trg):
+        input_shape = tf.shape(trg)
+        batch_size = input_shape[0]
+        seq_len = input_shape[1]
+        causal_mask = self.causal_attention_mask(batch_size, seq_len, seq_len, tf.bool)
+        trg_att = self.self_att(trg, trg, attention_mask=causal_mask)
+        trg_norm = self.ln1(trg + self.self_drop(trg_att))
+        enc_out = self.enc_att(trg_norm, enc_out)
+        enc_out_norm = self.ln2(self.enc_drop(enc_out) + trg_norm)
+        ffn_out = self.ffn(enc_out_norm)
+        ffn_out_norm = self.ln3(enc_out_norm + self.ffn_drop(ffn_out))
+        return ffn_out_norm
+
+
+"""
+## Complete Transformer Model
+
+Our model takes audio spectrograms as input, and predicts a sequence of characters.
+During training, we give the decoder the target character sequence shifted to the left
+as input. During inference, the decoder uses its own past predictions to predict the 
+next token.
+"""
+
+loss_object = tf.keras.losses.CategoricalCrossentropy(
+    from_logits=True, label_smoothing=0.1, reduction="none"
+)
+
+
+def masked_loss(real, pred):
+    """ assuming pad token index = 0 """
+    one_hot = tf.one_hot(real, depth=34)
+    mask = tf.math.logical_not(tf.math.equal(real, 0))
+    loss_ = loss_object(one_hot, pred)
+    mask = tf.cast(mask, dtype=loss_.dtype)
+    loss_ *= mask
+    return tf.reduce_mean(loss_)
+
+
+class Transformer(keras.Model):
+    def __init__(
+        self,
+        num_hid=64,
+        num_head=2,
+        num_ff=128,
+        src_maxlen=100,
+        trg_maxlen=100,
+        num_layers_enc=4,
+        num_layers_dec=1,
+        num_classes=10,
+    ):
+        super().__init__()
+        self.loss_metric = keras.metrics.Mean(name="loss")
+        self.num_layers_enc = num_layers_enc
+        self.num_layers_dec = num_layers_dec
+        self.trg_maxlen = trg_maxlen
+
+        self.enc_input = TransformerInput(
+            input_type="features", num_hid=num_hid, maxlen=src_maxlen
+        )
+        self.dec_input = TransformerInput(
+            input_type="tokens",
+            num_vocab=num_classes,
+            num_hid=num_hid,
+            maxlen=trg_maxlen,
+        )
+        for i in range(num_layers_enc):
```

Comment:
Use a Sequential (with a list comprehension) instead of setattr/getattr, it will be much simpler.

```
self.encoder = keras.Sequential(
   [PositionalEmbedding(...)] + [TransformerEncoder(...) for _ in range(num_encoder_layers)]
)
```
then you don't even the encoder/decoder methods, just do e.g. `self.encoder(inputs)`